youtube/feed.go

```diff
@@ -276,39 +277,36 @@ var (
 )
 
-func (p *Plugin) parseYtUrl(url string) (t ytUrlType, id string, err error) {
-	if ytUrlShortRegex.MatchString(url) {
-		capturingGroups := ytUrlShortRegex.FindAllStringSubmatch(url, -1)
-		if len(capturingGroups) > 0 && len(capturingGroups[0]) > 0 && len(capturingGroups[0][2]) > 0 {
-			return ytUrlTypeVideo, capturingGroups[0][2], nil
-		}
-	} else if ytVideoUrlRegex.MatchString(url) {
-		capturingGroups := ytVideoUrlRegex.FindAllStringSubmatch(url, -1)
-		if len(capturingGroups) > 0 && len(capturingGroups[0]) > 0 && len(capturingGroups[0][4]) > 0 {
-			return ytUrlTypeVideo, capturingGroups[0][4], nil
-		}
-	} else if ytChannelUrlRegex.MatchString(url) {
-		capturingGroups := ytChannelUrlRegex.FindAllStringSubmatch(url, -1)
-		if len(capturingGroups) > 0 && len(capturingGroups[0]) > 0 && len(capturingGroups[0][5]) > 0 {
-			return ytUrlTypeChannel, capturingGroups[0][5], nil
-		}
-	} else if ytCustomUrlRegex.MatchString(url) {
-		capturingGroups := ytCustomUrlRegex.FindAllStringSubmatch(url, -1)
-		if len(capturingGroups) > 0 && len(capturingGroups[0]) > 0 && len(capturingGroups[0][5]) > 0 {
-			return ytUrlTypeCustom, capturingGroups[0][5], nil
-		}
-	} else if ytUserUrlRegex.MatchString(url) {
-		capturingGroups := ytUserUrlRegex.FindAllStringSubmatch(url, -1)
-		if len(capturingGroups) > 0 && len(capturingGroups[0]) > 0 && len(capturingGroups[0][5]) > 0 {
-			return ytUrlTypeUser, capturingGroups[0][5], nil
-		}
-	} else if ytHandleUrlRegex.MatchString(url) {
-		capturingGroups := ytHandleUrlRegex.FindAllStringSubmatch(url, -1)
-		if len(capturingGroups) > 0 && len(capturingGroups[0]) > 0 && len(capturingGroups[0][5]) > 0 {
-			return ytUrlTypeHandle, capturingGroups[0][5], nil
-		}
+func (p *Plugin) parseYtUrl(url string) (idType ytUrlType, id string, err error) {
+	var valid bool
+	if id, valid = p.getIDFromRegexUrl(ytUrlShortRegex, url, 2); valid {
+		idType = ytUrlTypeVideo
+	} else if id, valid = p.getIDFromRegexUrl(ytVideoUrlRegex, url, 4); valid {
+		idType = ytUrlTypeVideo
+	} else if id, valid = p.getIDFromRegexUrl(ytChannelUrlRegex, url, 5); valid {
+		idType = ytUrlTypeChannel
+	} else if id, valid = p.getIDFromRegexUrl(ytCustomUrlRegex, url, 5); valid {
+		idType = ytUrlTypeCustom
+	} else if id, valid = p.getIDFromRegexUrl(ytUserUrlRegex, url, 5); valid {
+		idType = ytUrlTypeUser
+	} else if id, valid = p.getIDFromRegexUrl(ytHandleUrlRegex, url, 5); valid {
+		idType = ytUrlTypeHandle
+	} else {
+		idType = ytUrlTypeInvalid
+		err = errors.New("invalid or incomplete url")
 	}
-	return ytUrlTypeInvalid, "", errors.New("invalid or incomplete url")
+
+	return idType, id, err
+}
+
+func (p *Plugin) getIDFromRegexUrl(expr *regexp.Regexp, url string, index int) (id string, valid bool) {
+	capturingGroups := expr.FindAllStringSubmatch(url, -1)
+	if len(capturingGroups) > 0 && len(capturingGroups[0]) > index {
+		id = capturingGroups[0][index]
+		valid = id != ""
+	}
+
+	return
```

Comment:
Remove the naked return here as well — it adds no value.
```suggestion
	capturingGroups := expr.FindAllStringSubmatch(url, -1)
	if len(capturingGroups) > 0 && len(capturingGroups[0]) > index {
		return capturingGroups[0][index], true
	}
	return "", false
```

Reply:
This actually changes the existing behaviour, as the `if` statement succeeding doesn't result in `valid`/`ok` being implicitly true.

In this case, the named variables make the resulting `!= ""` check more concise.

Reply:
In the existing implementation, this was expressed as `len(capturingGroups[0][i]) > 0` instead, though these are evidently equivilent.

Reply:
The check is actually unnecessary as the regex patterns in question ensure that the relevant capturing groups will always be non-empty, but fair point. Deferring to @ashishjh-bst.

Reply:
The regex handles that, so this check will never be done here.